platforms/software/dependency-management/src/main/java/org/gradle/internal/component/model/LocalComponentDependencyMetadata.java

```diff
@@ -25,15 +25,22 @@
 import org.gradle.api.internal.attributes.AttributeContainerInternal;
 import org.gradle.api.internal.attributes.AttributesSchemaInternal;
 import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.internal.component.SelectionFailureHandler;
 import org.gradle.internal.deprecation.DeprecationLogger;
-import org.gradle.internal.exceptions.ConfigurationNotConsumableException;
 import org.gradle.util.internal.GUtil;
 
 import javax.annotation.Nullable;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 
+/**
+ * Information about a locally resolved dependency.
+ *
+ * An instance of {@link SelectionFailureHandler} is supplied to {@link #selectVariants(GraphVariantSelector, ImmutableAttributes, ComponentGraphResolveState, AttributesSchemaInternal, Collection)},
```

Comment:
🤔 Shouldn't this be documented on the `selectVariants` method instead of the top-level class? Perhaps even on `DependencyMetadata#selectVariants` since it applies to all dependency types

Reply:
I _do_ want some notice at the top-level that you should route failures through `SelectionFailureHandler` though, as it's tempting to start throwing selection-related exceptions directly in the implementations.  Perhaps as an `@implSpec` on the interface.